services/wireguard/handshake.go

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2020 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package wireguard
+
+import (
+	"context"
+	"errors"
+	"net"
+	"time"
+)
+
+// WaitHandshake waits until WireGuard does initial handshake.
+func WaitHandshake(statsFetch func() (*Stats, error), timeout time.Duration, stop chan struct{}) error {
```

Comment:
return a cancel func instead of passing in a stop.

Reply:
I can't return cancel func since there is a for loop which checks for latests handshake and returns err if any. It make sense to return cancel func for similar cases like context.WithTimeout. In this case stop channel can be use to get similar cleanup logic as with defer cancel.

```go
done := make(chan struct{})
defer close(done)
doSomething(done)
```